frame/vesting/src/tests.rs

```diff
@@ -307,53 +513,644 @@ fn force_vested_transfer_works() {
+		assert_eq!(Vesting::vesting(&2).unwrap(), vec![sched0, sched1]);
+
+		// Got to half way through the second schedule where both schedules are actively vesting
+		let cur_block = 20;
+		System::set_block_number(cur_block);
+
+		// Account 2 has no usable balances prior to the merge because they have not unlocked
+		// with `vest` yet.
+		assert_eq!(Balances::usable_balance(&2), 0);
+
+		assert_ok!(Vesting::merge_schedules(Some(2).into(), 0, 1));
+
+		// Merging schedules vests all pre-existing schedules prior to merging, which is reflected
+		// in account 2's updated usable balance.
+		let sched0_vested_now = sched0.per_block() * (cur_block - sched0.starting_block());
+		let sched1_vested_now = sched1.per_block() * (cur_block - sched1.starting_block());
+		assert_eq!(Balances::usable_balance(&2), sched0_vested_now + sched1_vested_now);
+
+		// The locked amount is the sum of what both schedules have locked at the current block.
+		let sched2_locked = sched1
+			.locked_at::<Identity>(cur_block)
+			.saturating_add(sched0.locked_at::<Identity>(cur_block));
+		// End block of the new schedule is the greater of either merged schedule.
+		let sched2_end = sched1
+			.ending_block_as_balance::<Identity>()
+			.max(sched0.ending_block_as_balance::<Identity>());
+		let sched2_duration = sched2_end - cur_block;
+		// Based off the new schedules total locked and its duration, we can calculate the
+		// amount to unlock per block.
+		let sched2_per_block = sched2_locked / sched2_duration;
+
+		let sched2 = VestingInfo::new(sched2_locked, sched2_per_block, cur_block);
+		assert_eq!(Vesting::vesting(&2).unwrap(), vec![sched2]);
+
+		// And just to double check, we assert the new merged schedule we be cleaned up as expected.
+		System::set_block_number(30);
+		vest_and_assert_no_vesting::<Test>(2);
+	});
+}
+
+#[test]
+fn merging_shifts_other_schedules_index() {
+	// Schedules being merged are filtered out, schedules to the right of any merged
+	// schedule shift left and the merged schedule is always last.
+	ExtBuilder::default().existential_deposit(ED).build().execute_with(|| {
+		let sched0 = VestingInfo::new(
+			ED * 10,
+			ED, // Vesting over 10 blocks
+			10,
+		);
+		let sched1 = VestingInfo::new(
+			ED * 11,
+			ED, // Vesting over 11 blocks
+			11,
+		);
+		let sched2 = VestingInfo::new(
+			ED * 12,
+			ED, // Vesting over 12 blocks
+			12,
+		);
+
+		// Account 3 starts out with no schedules,
+		assert_eq!(Vesting::vesting(&3), None);
+		// and some usable balance.
+		let usable_balance = Balances::usable_balance(&3);
+		assert_eq!(usable_balance, 30 * ED);
+
+		let cur_block = 1;
+		assert_eq!(System::block_number(), cur_block);
+
+		// Transfer the above 3 schedules to account 3.
+		assert_ok!(Vesting::vested_transfer(Some(4).into(), 3, sched0));
+		assert_ok!(Vesting::vested_transfer(Some(4).into(), 3, sched1));
+		assert_ok!(Vesting::vested_transfer(Some(4).into(), 3, sched2));
```

Comment:
This whole section can be reduced with some helpers, in most of the tests, see my comment above.

Reply:
https://github.com/paritytech/substrate/issues/9345